### PR TITLE
Implementing negative striding for python lists. Fix #25135

### LINKED ISSUE #25135: Negative step in index gives confusing error in JIT
For
```
def t():
    x = [1,2,3]
    x = x[::-1]
    print(x)

torch.jit.script(t)()
```
it gives `IndexError: vector::_M_range_check: __n (which is 18446744073709551615) >= this->size() (which is 3)`

@suo 

cc @ezyang @gchanan @zou3519 @bdhirsh @gmagogsfm @suo

### LINKED ISSUE #27543: [JIT] Reverse slicing List gives runtime error
## To Reproduce

```
@torch.jit.script
def fn():
    a = [1, 2, 3, 4]
    return a[::-1]
fn()
```

> RuntimeError: vector::_M_range_check The above operation failed in interpreter, with the following stack trace: at test/test_jit.py:6263:19
        @torch.jit.script
        def test():
            a = [1, 2, 3, 4]
            return a[::-1]
                   ~~~~~~ <--- HERE



This should either give a compile error, or work (better), but it should not compile and be unsupported.

cc @suo

### PR DESCRIPTION
**Pull request by agolynski**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#33019 Implementing negative striding for python lists. Fix #25135**

Differential Revision: [D19836031](https://our.internmc.facebook.com/intern/diff/D19836031)

### CODE REVIEW

File: torch/csrc/jit/script/compiler.cpp [hunk 1]
```diff
@@ -2965,6 +2962,7 @@ struct to_ir {
 
     auto step = emitExpr(Expr(slice.stepOr(1)));
     NamedValue step_nv = NamedValue(loc, "step", step);
+
```
Comment (albanD) [inline torch/csrc/jit/script/compiler.cpp]: nit: spurious new line

File: torch/csrc/jit/script/tree_views.h [hunk 2]
```diff
@@ -893,17 +893,18 @@ struct SliceExpr : public Expr {
...
-  Expr endOr(int alternative) const {
+  Expr endOr(int64_t alternative) const {
     const auto endOption = end();
     return endOption.present() ? endOption.get() : createInt(alternative);
   }
-  Expr stepOr(int alternative) const {
+  Expr stepOr(int64_t alternative) const {
     const auto stepOption = step();
-    return stepOption.present() ? stepOption.get() : createInt(alternative);
+    auto ret = stepOption.present() ? stepOption.get() : createInt(alternative);
```
Comment (albanD) [inline torch/csrc/jit/script/tree_views.h]: nit: What is the reason for this change? Just linting?

**Review by albanD — COMMENTED** [review]
lgtm
Do wait for a stamp for someone from the jit team though.

File: torch/csrc/jit/script/compiler.cpp (hunk 1 above)
Comment (agolynski) [reply torch/csrc/jit/script/compiler.cpp]: Done

File: torch/csrc/jit/script/tree_views.h (hunk 2 above)
Comment (agolynski) [reply torch/csrc/jit/script/tree_views.h]: Removed old debugging

**Comment by agolynski** [comment]
Added backwards compatibility overrride

File: aten/src/ATen/native/TensorShape.cpp [hunk 3]
```diff
@@ -706,6 +706,11 @@ Tensor slice(const Tensor& self, int64_t dim, int64_t start, int64_t end, int64_
   auto strides = self.strides().vec();
   // TODO: support negative strides
   TORCH_CHECK(step > 0, "slice step must be positive");
+
+  // INT64_MAX stands for default value.
+  if (start == INT64_MAX) {
```
Comment (eellison) [inline aten/src/ATen/native/TensorShape.cpp]: Why don't we have the same adjustment of start according to step as we have in lists ?
```
  if (*start == INT64_MAX) {
    *start = (step < 0) ? INT64_MAX : 0;
  }
```

Also don't we need to do the same adjustment of stop ?
```
  if (*stop == INT64_MAX) {
    *stop = (step < 0) ? INT64_MIN : INT64_MAX;
  }
```

**Review by eellison — COMMENTED** [review]
Thanks for the PR! Looks great! I have some small comments to make sure I'm understanding everything correctly.

File: torch/csrc/jit/register_prim_ops.cpp [hunk 4]
```diff
@@ -2580,9 +2623,11 @@ RegisterOperators reg2({
           "[] b) -> " decl_type "[]",                                               \
           listInplaceAdd<c_type::value_type>,                                       \
           aliasAnalysisFromSchema()),                                               \
+      /* INT64_MAX=9223372036854775807 represents unspeficied value, aka Py_None */ \
       Operator(                                                                     \
           "aten::slice(" decl_type                                                  \
-          "[] l, int start, int end=9223372036854775807, int step=1) -> " decl_type \
+          "[] l, int start=9223372036854775807,                                     \
```
[the remark this answers is not in the document]
Comment (agolynski) [reply torch/csrc/jit/register_prim_ops.cpp]: I found some peculiarities of how optional works in jit compiler, e.g. we make both ints optional and only second one is specified, e.g. [:42:], the implementation of slice receives the first one = 42 and second one being not specified for some reason. Also implementation receives ivalues so it has no way of checking names of params it receives.

File: torch/csrc/jit/register_prim_ops.cpp [hunk 5]
```diff
@@ -1975,35 +1975,78 @@ int listMulIntRight(Stack& stack) {
...
 
+// Stolen (with appropriate modifications) from cpython repo
+// Objects/sliceobject.c with comment:
+// this is harder to get right than you might think
+//
+// This adjusts indexes according to python list semantics and returns number
+// of elements in the resulting list.
+static int64_t PySlice_AdjustIndices(
+    int64_t length, int64_t* start, int64_t* stop, int64_t step) {
+  TORCH_CHECK(step != 0, "List slice should have non-zero step")
```
[the remark this answers is not in the document]
Comment (agolynski) [reply torch/csrc/jit/register_prim_ops.cpp]: I simply didn't quite understand what they meant by 
assert(step >= -PY_SSIZE_T_MAX);
and why do we need the check in our implementation.

I believe there is already int64 boundaries checks when passing numbers from python to cpp?

File: torch/csrc/jit/register_prim_ops.cpp [hunk 6]
```diff
@@ -1975,35 +1975,78 @@ int listMulIntRight(Stack& stack) {
...
+
+  if (*start == INT64_MAX) {
+    *start = (step < 0) ? INT64_MAX : 0;
+  }
+  if (*stop == INT64_MAX) {
+    *stop = (step < 0) ? INT64_MIN : INT64_MAX;
+  }
```
[the remark this answers is not in the document]
Comment (agolynski) [reply torch/csrc/jit/register_prim_ops.cpp]: Done

File: torch/csrc/jit/register_prim_ops.cpp [hunk 7]
```diff
@@ -1975,35 +1975,78 @@ int listMulIntRight(Stack& stack) {
...
-  const auto normalized_end =
-      std::min(list_size, normalizeIndex(end, list_size));
-
   c10::List<T> sliced_list = make_result_list<T>(list.elementType());
-  if (normalized_end <= normalized_start) {
-    // early exit if the slice is trivially empty
+  const int64_t num_values =
+       PySlice_AdjustIndices(list_size, &start, &stop, step);
+
+  if (num_values == 0) {
```
[the remark this answers is not in the document]
Comment (agolynski) [reply torch/csrc/jit/register_prim_ops.cpp]: agreed (I just kept the existing code structure). I removed it now.

File: aten/src/ATen/native/TensorShape.cpp (hunk 3 above)
Comment (agolynski) [reply aten/src/ATen/native/TensorShape.cpp]: Stop being maxint is a preexisting condition. Since we only support step > 0 here, stop should be INT64_MAX if not specified.

I've asked Edward if he'd like me to undertake supporting step < 0 here, but seems like the project is more complicated than just changing impl like it was done for lists. Negative tensor striding has to be supported on native level - we can't simply copy values here.

File: torch/csrc/jit/register_prim_ops.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (albanD) [reply torch/csrc/jit/register_prim_ops.cpp]: The point of this test is to prevent `step == INT64_MIN`. Because in that case `-step = step` (there is one more negative value than positive value and this - will overflow back to INT64_MIN) and line 2015 below will not work as expected.
I think you want to add it back, stating that a step of `INT64_MIN` is not supported (and hopefully no one will ever see that error message! )

[the remark this answers is not in the document]
Comment (agolynski) [reply torch/csrc/jit/register_prim_ops.cpp]: Done and added tests.

As we've looked with Alban, the overall behavior is actually correct without this check, but let's be safe and compatible with cpython :)

File: test/test_jit.py [hunk 8]
```diff
@@ -5977,6 +5991,33 @@ def func(x, value1, value2):
...
+        check_indexing_list_int('[-1]', consec_list(10))
+        check_indexing_list_int('[0:2]', consec_list(11))
+        check_indexing_list_int('[1:-1]', consec_list(12))
+        check_indexing_list_int('[-3:-1]', consec_list(13))
+        check_indexing_list_int('[1:]', consec_list(15))
+        check_indexing_list_int('[:1]', consec_list(16))
+        check_indexing_list_int('[:]', consec_list(17))
+        check_indexing_list_int('[::]', consec_list(0))
+        check_indexing_list_int('[1000::]', consec_list(0))
+        check_indexing_list_int('[:1000:]', consec_list(0))
```
[the remark this answers is not in the document]
Comment (eellison) [reply test/test_jit.py]: Oops, i meant the case `[:42:]` you mentioned below.

File: torch/csrc/jit/register_prim_ops.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (eellison) [reply torch/csrc/jit/register_prim_ops.cpp]: so the lexing of slice needs to be updated so that `:` correctly marks the `start` as None ?

**Review by eellison — COMMENTED** [review]
Everything you've added is good, but as you mentioned we need to add handling of `:42:` case

**Review by eellison — DISMISSED** [review]
LGTM, ship it !

                   v  ~.      v
          v           /|
                     / |          v
              v     /__|__
                  \--------/

**Comment by agolynski** [comment]
Thank you for a thorough review! Landing..

**Comment by agolynski** [comment]
Sigh... this breaks tuple slicing (due to start param having a different default). I'll update the PR and ping this again.

https://circleci.com/gh/pytorch/pytorch/4462358?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

and code is here:

https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/register_prim_ops.cpp#L988

File: torch/csrc/jit/script/compiler.cpp [hunk 9]
```diff
@@ -2937,17 +2937,14 @@ struct to_ir {
...
     if (dim) {
       AT_ASSERT(input->type()->isSubtypeOf(TensorType::get()));
-
       args.emplace_back(dim);
     } else {
       AT_ASSERT(!input->type()->isSubtypeOf(TensorType::get()));
     }
 
-    args.emplace_back(loc, "begin", emitExpr(Expr(slice.startOr(0))));
+    args.emplace_back(loc, "start", emitExpr(Expr(slice.startOr(INT64_MAX))));
```
Comment (eellison) [inline torch/csrc/jit/script/compiler.cpp]: Can you just move the logic for default `start` and `end` within the TupleSlice? Because you've added defaults to the schema declaration in both list slicing and tensor slicing the default logic in the compiler is only necessary for the tuple slicing logic. There are also downstream uses of the indices in `lower_tuples.cpp` uses that (reasonably) assume that a start of 0 means that that the starting index is 0.  

It's a little unfortunate that we have a divergence between tuple slicing and tensor slicing, but that divergence already existed and I don't think it's worth reconciling.

**Review by eellison — COMMENTED** [review]
To summarize off-line discussion, plan of action for moving this forward:

- Change declaration of `aten::slice` to
`func: slice.Tensor(Tensor(a) self, int dim=0, int? start=None, int? end=None, int step=1) -> Tensor(a)`
- since this is a forward-incompatible (but not backwards incompatible change), it will require some care landing. 

To land:
Land Part 1:
Land all of your changes except the changes to the compiler. Switch the schema to take optionals, keep the defaults as they are (0, MAX_INT). Still implement the handling of when they are None. 

Land Part 2:
2. Make the changes to the compiler, and update the defaults to be None.

File: torch/csrc/jit/register_prim_ops.cpp [hunk 10]
```diff
@@ -1975,35 +1975,75 @@ int listMulIntRight(Stack& stack) {
   return 0;
 }
 
+// Stolen (with appropriate modifications) from cpython repo
```
Comment (apaszke) [inline torch/csrc/jit/register_prim_ops.cpp]: I think you might need to look into [Python's license](https://github.com/python/cpython/blob/master/LICENSE) before we can copy the code just like that. There are usually some restrictions like "you have to include this text along the sources".

File: torch/csrc/jit/register_prim_ops.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (apaszke) [reply torch/csrc/jit/register_prim_ops.cpp]: Aren't we already breaking BC due to the default argument change? IMHO we should make them all optional now and avoid the issues in the future.

[the remark this answers is not in the document]
Comment (eellison) [reply torch/csrc/jit/register_prim_ops.cpp]: Yea, that's the plan https://github.com/pytorch/pytorch/pull/33019#pullrequestreview-368203616. 

I don't think this actually breaks BC though.

File: torch/csrc/jit/register_prim_ops.cpp (hunk 10 above)
Comment (eellison) [reply torch/csrc/jit/register_prim_ops.cpp]: Looks like you can add `"Copyright (c) 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010,
2011, 2012, 2013, 2014, 2015, 2016, 2017, 2018, 2019, 2020 Python Software Foundation;
All Rights Reserved"` above the code and that is sufficient.
